Lib/unittest/case.py

```diff
@@ -153,6 +154,35 @@ def _is_subtype(expected, basetype):
         return all(_is_subtype(e, basetype) for e in expected)
     return isinstance(expected, type) and issubclass(expected, basetype)
 
+
+def _heuristic_diff(a: list[str], b: list[str]) -> Iterator[str]:
+    """After testing the magnitude of the inputs, preferably return the output
+    of difflib.ndiff, but fallback to difflib.unified_diff for prohibitively
+    expensive inputs. How cost is calclated:
+
+        cost = (number of differing lines
+                * total length of all differing lines)
+
+    See bpo-19217 for context.
+    """
+    cost_limit = 1_000_000
+
+    # unified diff is always cheap, so we can use it to measure the magnitude
+    # of the differences, which is a proxy for the cost of difflib.ndiff
+    udiff = list(difflib.unified_diff(a, b, fromfile="expected", tofile="got"))
+    udiff_differing_lines = [l for l in udiff
+                             if l.startswith('-') or l.startswith('+')]
```

Comment:
```suggestion
                             if l.startswith(('-', '+'))]
```
And then the whole listcomp fits on one line.